doc/man3/EVP_PKEY_sign.pod

```diff
@@ -26,37 +35,121 @@ for more information about implicit fetches.
+Passing the message is supported both in a one-shot fashion using
+EVP_PKEY_sign(), and through the update fashion functions
+EVP_PKEY_sign_message_update() and EVP_PKEY_sign_message_final().
+This function enables using algorithms that can process infinite input, such
+as ED25519, RSA-SHA256 and similar.
+
+EVP_PKEY_sign_message_update() adds I<inlen> bytes from I<in> to the data to be
+processed for signature.  The signature algorithm specification and
+implementation determine how the input bytes are processed and if there's a
+limit on the total size of the input.  See L</NOTES> below for a deeper
+explanation.
+
+EVP_PKEY_sign_message_final() signs the processed data and places the data in
+I<sig>, and the number of signature bytes in I<*siglen>, if the number of
+bytes doesn't surpass the size given by I<sigsize>.
+I<sig> may be NULL, and in that case, only I<*siglen> is updated with the
+number of signature bytes.
+
+EVP_PKEY_sign() is a one-shot function that can be used with all the init
+functions above.
+When initialization was done with EVP_PKEY_sign_init(), EVP_PKEY_sign_init_ex()
+of EVP_PKEY_sign_init_ex2(), the data specified by I<tbs> and I<tbslen> is
+signed after appropriate padding.
+When initialization was done with EVP_PKEY_sign_message_init(), the data
+specified by I<tbs> and I<tbslen> is digested by the implied message digest
+algorithm, and the result is signed after appropriate padding.
+If I<sig> is NULL then the maximum size of the output buffer is written to the
+I<siglen> parameter.
+If I<sig> is not NULL, then before the call the I<siglen> parameter should
+contain the length of the I<sig> buffer, and if the call is successful the
+signature is written to I<sig> and the amount of data written to I<siglen>.
 
 =head1 NOTES
 
-EVP_PKEY_sign() does not hash the data to be signed, and therefore is
-normally used to sign digests. For signing arbitrary messages, see the
-L<EVP_DigestSignInit(3)> and
-L<EVP_SignInit(3)> signing interfaces instead.
-
-After the call to EVP_PKEY_sign_init() algorithm specific control
-operations can be performed to set any appropriate parameters for the
-operation (see L<EVP_PKEY_CTX_ctrl(3)>).
-
-The function EVP_PKEY_sign() can be called more than once on the same
-context if several operations are performed using the same parameters.
+=begin comment
+
+These notes are largely replicated in EVP_PKEY_verify.pod, please keep them
+in sync.
+
+=end comment
+
+=head2 General
+
+Some signature implementations only accumulate the input data and do no
+further processing before signing it (they expect the input to be a digest),
+while others compress the data, typically by internally producing a digest,
+and signing the result.
+Some of them support both modes of operation at the same time.
+The caller is expected to know how the chosen algorithm is supposed to behave
+and under what conditions.
+
+For example, an RSA implementation can be expected to only expect a digest as
```

Comment:
digested message

Reply:
Mm, I did something slightly different, hope that's good enough